### ISSUE TITLE
create a build pack for the serverless.com framework

### ISSUE THREAD

**Issue by jstrachan** [jenkins-x/jx#553]
we should support projects using the serverless framework by detecting a `serverless.yml`

here's an example with kubeless with or without serverless
https://github.com/kubeless/functions/tree/master/incubator/twitter

so we'd want a `serverless` folder in the build pack repository: https://github.com/jenkins-x-buildpacks/jenkins-x-kubernetes/tree/master/packs which defines a PR and release pipeline for using the serverless.com framework to build/deploy versions of a serverless project using the `serverless` CLI tool in a container image.

Once we have that folks can import their projects with a `jenkins-x.yml` file of 

```yaml
buildPack: serverless
```

and their CI/CD should just work. 

Then it'd be nice to modify `jx import` so that it can detect the `serverless.yml` file and default the `jenkins-x.yml` file as above if there's not already a file existing.

**Comment by jstrachan** [comment]
relates to #554

**Comment by michaelneale** [comment]
could also have a quickstart?

**Comment by michaelneale** [comment]
would this install kubeless on the cluster if it wasn't already there? or would that be part of the initial setup?

**Comment by jstrachan** [comment]
Yeah we could do with a build pack & a QuickStart.

We could add an optional addon too (folks can install kubelese themselves from a helm chart if they want) but we may as well make it really easy to use Jenkins X and kubeless.

Eg `jx create addon kubeless` might be nice

**Comment by jstrachan** [comment]
yeah - I guess the QuickStart could have a chart with a dependency - though haven’t checked if 2 quickstarts depend on the same chart does helm only install it once?

so we may need another mechanism - eg to ensure dependencies are in the Environment `requirements.yaml` when doing a promotion. We could maybe add an `environmentRequirements.yaml` if needed that we could use in promotions

**Comment by michaelneale** [comment]
Right - would only want it once. Kubeless also wants to use a dedicated “kubeless” namespace so I wonder if there are any RBAC issues. In any case it could fall back with instructions on how to prepare it (or JenkinsX could ship with kubeless already in the box, but it does seem like an ideal add on)

**Comment by jstrachan** [comment]
I guess the pipeline could also verify the addon has been installed & fail the pipeline nicely with a useful error message (e.g. if someone imports a kubeless based quickstart into a Jenkins X cluster that doesn't have the addon)?

**Comment by jstrachan** [comment]
raised https://github.com/jenkins-x/jx/issues/584

another idea is - the `jenkins-x.yml` file could add the required addons needed; which the `jx import / jx create quickstart` commands could detect and prompt the user to install those addons?

**Comment by jstrachan** [comment]
also https://github.com/jenkins-x/jx/issues/585

**Comment by jstrachan** [comment]
@michaelneale ok #584 is fixed so a `kubeless` quickstart can include a `jenkins-x.yml` file like this:
https://github.com/jenkins-x/jx/blob/master/pkg/jx/cmd/test_data/project_with_kubeless/jenkins-x.yml
which will describe the fact that the project requires the `kubeless` addon to be installed.

Then the `Jenkinsfile` can include this command:
```
jx step validate
```
which will fail the pipeline if `kubeless` is not installed and give the user information for how to install the missing addon

**Comment by michaelneale** [comment]
I think the jenkins-x yml idea is ok - but wary of it if it lives around in the final repo (ie there is a pipeline in there, helm charts and more, so it may look a tad confusing to have something else?)

**Comment by jstrachan** [comment]
We're using it for typesafe yaml configuration of issue tracking configuration & other repo specific stuff - so using it to add addons along with optional config (e.g. minimum versions) sounds fine to me. 

A `Jenkinsfile` isn't a terribly good place to put easy to parse/validate/edit YAML yet. I guess another option is a YAML representation a `Jenkinsfile`? :)

Each tool needs its own yaml; `draft, skaffold, make, jenkins` so its only fair `jx` gets a config file too right?

**Comment by jstrachan** [comment]
We could rename to `.jenkins-x.yml` to try hide it?

**Comment by michaelneale** [comment]
@jstrachan if it is just used at quickstart/import time - that is probably fine. Otherwise, the import could exclude it before pushing it somehow? (just a thought).

**Comment by polothy** [comment]
@jstrachan just curious on your thoughts on why kubeless over something like openfaas?  I don't have any opinions on either of them at the moment, just curious on your reasoning.

**Comment by michaelneale** [comment]
This is a bit out of date now. Knative is offering most of this and I mostly already works. I am not sure what the faas layer of it is though: there was a knative faas by pivotal but then I also heard openfaas May make use of it.

**Comment by jstrachan** [comment]
docs on our knative support here: https://jenkins-x.io/developing/serverless-apps/

the main reason I plumbed for kubeless when I raised this issue a year ago was its a provider in serverless.com: https://serverless.com/framework/docs/providers/ as is fn - though doesn't look like openfaas is supported yet.

I'm still keen to support CI/CD for serverless.com so folks can deploy to Lambda/Googlle Cloud Functions/Azure Functions.

TBH for the k8s based solutions; we can probably support kubeless+openfaas via CRDs in the helm charts like we did for knative - a quickstart would probably do the trick; we probably don't need much extra work. e.g. we could do a k8s native kubeless / openfaas via vanilla k8s without needing serverless.com
